src/algorithms/preprune.rs

```diff
@@ -0,0 +1,116 @@
+    fn guess(&mut self, history: &[Guess]) -> String {
+        if let Some(last) = history.last() {
+            if matches!(self.remaining, Cow::Owned(_)) {
+                self.remaining
+                    .to_mut()
+                    .retain(|(word, _)| last.matches(word));
+            } else {
+                self.remaining = Cow::Owned(
+                    self.remaining
+                        .iter()
+                        .filter(|(word, _)| last.matches(word))
+                        .copied()
+                        .collect(),
+                );
+            }
+
+            // prune patterns
+            if matches!(self.patterns, Cow::Owned(_)) {
+                self.patterns
+                    .to_mut()
+                    .retain(|p| last.compatible_pattern(p));
+            } else {
+                self.patterns = Cow::Owned(
+                    self.patterns
+                        .iter()
+                        .copied()
+                        .filter(|p| last.compatible_pattern(p))
+                        .collect(),
+                );
+            }
+        }
+        if history.is_empty() {
+            return "tares".to_string();
+        }
+
+        let remaining_count: usize = self.remaining.iter().map(|&(_, c)| c).sum();
+
+        let mut best: Option<Candidate> = None;
+        for &(word, count) in &*self.remaining {
+            let mut sum = 0.0;
+            let mut self_total_count = 0usize;
+            for pattern in &*self.patterns {
+                // considering a world where we _did_ guess `word` and got `pattern` as the
+                // correctness. now, compute what _then_ is left.
+                let mut in_pattern_total = 0;
+                for (candidate, count) in &*self.remaining {
+                    let g = Guess {
+                        word: Cow::Borrowed(word),
+                        mask: *pattern,
+                    };
+                    if g.matches(candidate) {
+                        in_pattern_total += count;
+                    }
+                }
+                if in_pattern_total == 0 {
+                    continue;
+                }
+                self_total_count += in_pattern_total;
+
+                // TODO: apply sigmoid
+                let p_of_this_pattern = in_pattern_total as f64 / remaining_count as f64;
+                sum += p_of_this_pattern * p_of_this_pattern.log2();
+            }
+
+            debug_assert_eq!(self_total_count, remaining_count, "{}", word);
+
+            let p_word = count as f64 / remaining_count as f64;
+            let goodness = p_word * -sum;
+            if let Some(c) = best {
+                // Is this one better?
+                if goodness > c.goodness {
+                    best = Some(Candidate { word, goodness });
+                }
+            } else {
+                best = Some(Candidate { word, goodness });
+            }
+        }
+        best.unwrap().word.to_string()
```

Comment:
And here, before returning `best.word`, walk `remaining_patterns` (restored from the old `Prune`) and remove any whose count in `best.matches` is 0. That should give you a _correct_ pruning of `remaining_patterns` since it prunes any pattern that had no matching words _for the guess we ended up making_.